chapters/errors.Rmd

```diff
@@ -815,6 +881,111 @@ the default name of `collate.log`.
 2.  Modify your first solution to handle nonexistent files
     and permission problems separately.
 
+### Testing error handling {#errors-ex-error-handling}
+
+In Chapter \@ref(errors), we modified `collate.py` to handle
+different types of errors associated with reading input files.
+The relevant code appears in `main`:
+
+```python
+"""
+Combine multiple word count CSV-files
+into a single cumulative count.
+"""
+
+import csv
+import argparse
+from collections import Counter
+import logging
+
+import utilities as util
+
+
+def update_counts(reader, word_counts):
+    """Update word counts with data from another reader/file."""
+    for word, count in csv.reader(reader):
+        word_counts[word] += int(count)
+
+
+def main(args):
+    """Run the command line program."""
+    log_lev = logging.DEBUG if args.verbose else logging.WARNING
+    logging.basicConfig(level=log_lev, filename=args.logfile)
+    word_counts = Counter()
+    logging.info('Processing files...')
+    for fname in args.infiles:
+        try:
+            logging.debug(f'Reading in {fname}...')
+            if fname[-4:] != '.csv':
+                msg = util.ERRORS['not_csv_suffix'].format(
+                      fname=fname)
+                raise OSError(msg)
+            with open(fname, 'r') as reader:
+                logging.debug('Computing word counts...')
+                update_counts(reader, word_counts)
+        except FileNotFoundError:
+            msg = f'{fname} not processed: File does not exist'
+            logging.warning(msg)
+        except PermissionError:
+            msg = f'{fname} not processed: No permission to read'
+            logging.warning(msg)
+        except Exception as error:
+            msg = f'{fname} not processed: {error}'
+            logging.warning(msg)
+    util.collection_to_csv(word_counts, num=args.num)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument('infiles', type=str, nargs='*',
+                        help='Input file names')
+    parser.add_argument('-n', '--num', type=int, default=None,
+                        help='Output only n most frequent words')
+    parser.add_argument('-v', '--verbose',
+                        action="store_true", default=False,
+                        help="Set logging level to DEBUG")
+    parser.add_argument('-l', '--logfile',
+                        type=str, default='collate.log',
+                        help='Name of the log file')
+    args = parser.parse_args()
+    main(args)
+```
+
+In this chapter we discussed using `pytest.raises`
+to test whether the error handling in a program is working as expected (Section \@ref(errors-testing)).
```

Comment:
```suggestion

```

Reply:
I think it reads better if I take out this whole para.